data/src/main/java/org/apache/iceberg/data/TableMigrationUtil.java

```diff
@@ -141,28 +142,36 @@ public static List<DataFile> listPartition(
             index -> {
               Metrics metrics = getAvroMetrics(fileStatus.get(index).getPath(), conf);
               datafiles[index] =
-                  buildDataFile(fileStatus.get(index), partitionValues, spec, metrics, "avro");
+                  metrics.recordCount() > 0
+                      ? buildDataFile(fileStatus.get(index), partitionValues, spec, metrics, "avro")
+                      : null;
             });
       } else if (format.contains("parquet")) {
         task.run(
             index -> {
               Metrics metrics =
                   getParquetMetrics(fileStatus.get(index).getPath(), conf, metricsSpec, mapping);
               datafiles[index] =
-                  buildDataFile(fileStatus.get(index), partitionValues, spec, metrics, "parquet");
+                  metrics.recordCount() > 0
+                      ? buildDataFile(
+                          fileStatus.get(index), partitionValues, spec, metrics, "parquet")
+                      : null;
             });
       } else if (format.contains("orc")) {
         task.run(
             index -> {
               Metrics metrics =
                   getOrcMetrics(fileStatus.get(index).getPath(), conf, metricsSpec, mapping);
               datafiles[index] =
-                  buildDataFile(fileStatus.get(index), partitionValues, spec, metrics, "orc");
+                  metrics.recordCount() > 0
+                      ? buildDataFile(fileStatus.get(index), partitionValues, spec, metrics, "orc")
+                      : null;
             });
       } else {
         throw new UnsupportedOperationException("Unknown partition format: " + format);
       }
-      return Arrays.asList(datafiles);
+      return Arrays.asList(
+          Arrays.stream(datafiles).filter(Objects::nonNull).toArray(DataFile[]::new));
```

Comment:
Is it possible to avoid changes above and filter out empty files here? It will lead to have less conflicts during rebase/ cherry-picks.

Reply:
Something like this https://github.com/SinghAsDev/iceberg/commit/a0c998c2cdd01fc5c523428c59713e74e6071115